java/src/main/java/io/cucumber/cucumberexpressions/ExpressionFactory.java

```diff
@@ -29,14 +27,25 @@ public ExpressionFactory(ParameterTypeRegistry parameterTypeRegistry) {
     }
 
     public Expression createExpression(String expressionString) {
-        if (BEGIN_ANCHOR.matcher(expressionString).find() || END_ANCHOR.matcher(expressionString).find()) {
-            return createRegularExpressionWithAnchors(expressionString);
+        int length = expressionString.length();
+        int lastCharIndex = 0;
+        char firstChar = 0;
+        char lastChar = 0;
+        if (length > 0) {
+            lastCharIndex = length - 1;
+            firstChar = expressionString.charAt(0);
+            lastChar = expressionString.charAt(lastCharIndex);
         }
-        Matcher m = SCRIPT_STYLE_REGEXP.matcher(expressionString);
-        if (m.find()) {
-            return new RegularExpression(Pattern.compile(m.group(1)), parameterTypeRegistry);
+        if (firstChar == '^' || lastChar == '$') {
+            // java regexp => create from regexp
```

Comment:
This and the other added comments can be removed. I was looking for comments to explain what has been optimized and _why_ it was more optimal so people will be careful with refactoring in the future The added comments mostly describe a _what_.